### PR TITLE
boards: nrf9160dk: Add support for newer revisions (0.14.0+) of the DK

### PR DESCRIPTION
**Pull request by anangl**
- Use devicetree instead of Kconfig to configure the board control switches in nRF9160 DK:
  - add binding for the switches that provide optional signal routings on this board
  - add binding for the GPIO interface that can be used for communication (e.g. UART based) between the nRF9160 and the nRF52840 on the DK, and add GPIO mapping for this interface so that its lines can be used without caring about of actual pin numbers on both sides
  - add binding for one GPIO line chosen from the above interface that is to allow the nRF9160 to reset the nRF52840
  - update accordingly dts files and board specific code for both board definitions associated with the DK
  - introduce .dtsi files that can be included from dts overlays in order to facilitate the use of the above GPIO interface; modify the overlay in the hci_uart sample to provide an example of use of those files

- Use the multiple board revisions feature to provide support for the new hardware possibilities available in nRF9160 DK starting from v0.14.0.

- Add DT node and .dtsi files that allow using the PCAL6408A I/O expander that is available in nRF9160 DK v0.14.0 or later to interface LEDs and buttons on the DK

### CODE REVIEW

**Review by galak — COMMENTED** [review]
Feels like BOARD_REVISION/BOARD_REVISION_STRING need documentation somewhere.

**Comment by tejlmand** [comment]
> Feels like BOARD_REVISION/BOARD_REVISION_STRING need documentation somewhere.

If you consider the board revision feature itself, then that is documented here:
https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#multiple-board-revisions

If you refer to this commit: https://github.com/zephyrproject-rtos/zephyr/pull/31624/commits/f817ff32e4993ddc733094753ae3193b0119959e
then there is a dedicated PR for that here: https://github.com/zephyrproject-rtos/zephyr/pull/31617

If you mean you would like to see dedicated documentation for those, then it's probably outside the scope of this PR, but I agree that we have too limited documentation for Zephyr / CMake in general.
Updated https://github.com/zephyrproject-rtos/zephyr/issues/9947 with additional content.

**Comment by anangl** [comment]
Rebased on #31864.

**Comment by anangl** [comment]
Rebased on master after #31864 has been merged.

**Comment by anangl** [comment]
Rebased on #31926.

File: boards/arm/nrf9160dk_nrf52840/board.c
Comment (mbolivar-nordic) [inline boards/arm/nrf9160dk_nrf52840/board.c:15]: Might want to call this one GET_CTLR instead, for consistency.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (mbolivar-nordic) [inline boards/arm/nrf9160dk_nrf52840/doc/index.rst:170]: Did you mean "the given routing" or something like that?

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (mbolivar-nordic) [inline boards/arm/nrf9160dk_nrf52840/doc/index.rst:195]: Please use "node label" instead of "node alias", since these are not in `/aliases`.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (mbolivar-nordic) [inline boards/arm/nrf9160dk_nrf52840/doc/index.rst:197]: Did you consider naming these to match the analog switch name more closely, e.g. "nrf91_uart1" instead of "vcom0_pins_routing", "nrf91_led1" instead of "led1_pin_routing", etc.? If so, why did you decide against it?

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (mbolivar-nordic) [inline boards/arm/nrf9160dk_nrf52840/doc/index.rst:243]: Consider using `:zephyr_file:` to link directly to these helpers.

File: boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml
Comment (mbolivar-nordic) [inline boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml:4]: I think it should be possible to use pre_dt_board.cmake from the nrf9160dk_nrf52840 board directory to manipulate DTS_ROOT. By adding the nrf9160dk_nrf9160 board directory to DTS_ROOT, this binding should be available without copy/pasting.

**Review by mbolivar-nordic — COMMENTED** [review]
Nice work!

I need a bit more time with this PR to work through the details, but I have a few minor things from a first pass.

File: boards/arm/nrf9160dk_nrf52840/board.c
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/board.c]: Of course, that was my intention. I just made a typo here.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/doc/index.rst]: Yes. Corrected.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/doc/index.rst]: Sure, I meant the node label but I used a wrong term.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/doc/index.rst:197]: I thought about it but I was afraid it could be confusing. Especially for the nRF91_UART1 switch which in fact provides routing for `uart0` in the nrf9160dk_nrf9160 board. I wanted to provide more meaningful names as node labels. The nodes themselves are named exactly after the names of analog switches from the [DK documentation](https://infocenter.nordicsemi.com/topic/ug_nrf91_dk/UG/nrf91_DK/board_controller.html). The above node is named `switch-nrf91-uart1` and in the generated `zephyr.dts` file the user will see:
```
		vcom0_pins_routing: switch-nrf91-uart1 {
			compatible = "nordic,nrf9160dk-optional-routing";
			control-gpios = < &gpio1 0xe 0x1 >;
			status = "okay";
		};
```
I think it should be clear enough.
But I'm open to change the naming of those nodes labels if the current one is actually confusing or inconvenient.

File: boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml:4]: I'm not sure how to do it properly, as I don't want the other files from `dts/` directory of one board to be available for the other. Only the two files: `nordic,nrf9160dk-nrf52840-interface.yaml` and `nordic,nrf9160dk-nrf52840-reset.yaml` are common for both these boards. Should I move these two files to another directory and add it to DTS_ROOT?

File: boards/arm/nrf9160dk_nrf9160/doc/index.rst
Comment (ru-fu) [inline boards/arm/nrf9160dk_nrf9160/doc/index.rst:131]:
```suggestion
for :ref:`nrf9160dk_nrf52840` (see the **Board controller firmware** section
```

File: boards/arm/nrf9160dk_nrf9160/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf9160/doc/index.rst:86]: Moved.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/doc/index.rst]: I rephrased this note a bit and placed it earlier in the section (right after the first paragraph).

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/doc/index.rst]: Moved.

File: boards/arm/nrf9160dk_nrf52840/doc/index.rst
Comment (anangl) [reply boards/arm/nrf9160dk_nrf52840/doc/index.rst]: I chose the first option, as nRF52840 SoC pins are something similar to those UART interfaces of the interface MCU, so I think they should be regarded as a DK component. :)

**Comment by anangl** [comment]
Rebased on master after #31926 has been merged.

**Comment by anangl** [comment]
Rebased to solve a merge conflict.

**Comment by anangl** [comment]
Rebased to make CI green.

File: boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml
Comment (jfischer-no) [reply boards/arm/nrf9160dk_nrf52840/dts/bindings/nordic,nrf9160dk-nrf52840-interface.yaml:4]: Is there anything against having only nrf9160dk directory for both boards nrf9160dk_nrf9160 and nrf9160dk_nrf52840?

File: drivers/gpio/CMakeLists.txt
Comment (jfischer-no) [inline drivers/gpio/CMakeLists.txt:38]: I suggest to open separate PR for PCAL6408A driver.

Comment (anangl) [reply drivers/gpio/CMakeLists.txt:38]: Done. See #32748.

**Comment by anangl** [comment]
Rebased on #32748.

**Review by mbolivar-nordic — APPROVED** [review]
Lgtm once unblocked.

**Comment by anangl** [comment]
Rebased after #32748 has been merged.

**Review by jfischer-no — APPROVED** [review]

**Review by ru-fu — APPROVED** [review]
Docs look good.

**Review by ioannisg — APPROVED** [review]

**Pull request merged**
